src/frontend/Typechecker.ml

```diff
@@ -0,0 +1,1549 @@
+
+  Functions which begin with "check_" return a typed version of their input
+  Functions which begin with "verify_" return unit if a check succeeds, or else
+    throw an Errors.SemanticError exception.
+
+  All Error.SemanticError excpetions are caught by check_program
+  which turns the ast or exception into a Result.t for external usage
+
+  A type environment (Env.t) is used to hold variables and functions, including
+  stan math functions. This is a functional map, meaning it is handled immutably.
+*)
+
+open Core_kernel
+open Middle
+open Ast
+module Env = Environment
+
+(* we only allow errors raised by this function *)
+let error e = raise (Errors.SemanticError e)
+
+(* model name - don't love this here *)
+let model_name = ref ""
+let check_that_all_functions_have_definition = ref true
+
+(* Record structure holding flags and other markers about context to be
+   used for error reporting. *)
+type context_flags_record =
+  { current_block: Env.originblock
+  ; in_toplevel_decl: bool
+  ; in_fun_def: bool
+  ; in_returning_fun_def: bool
+  ; in_rng_fun_def: bool
+  ; in_lp_fun_def: bool
+  ; in_udf_dist_def: bool
+  ; loop_depth: int }
+
+let context block =
+  { current_block= block
+  ; in_toplevel_decl= false
+  ; in_fun_def= false
+  ; in_returning_fun_def= false
+  ; in_rng_fun_def= false
+  ; in_lp_fun_def= false
+  ; in_udf_dist_def= false
+  ; loop_depth= 0 }
+
+let calculate_autodifftype cf at ut =
+  match at with
+  | Env.(Param | TParam | Model | Functions)
+    when not (UnsizedType.contains_int ut || cf.current_block = GQuant) ->
+      UnsizedType.AutoDiffable
+  | _ -> DataOnly
+
+let arg_type x = (x.emeta.ad_level, x.emeta.type_)
+let get_arg_types = List.map ~f:arg_type
+let type_of_expr_typed ue = ue.emeta.type_
+let has_int_type ue = ue.emeta.type_ = UInt
+let has_int_array_type ue = ue.emeta.type_ = UArray UInt
+
+let has_int_or_real_type ue =
+  match ue.emeta.type_ with UInt | UReal -> true | _ -> false
+
+let probability_distribution_name_variants id =
+  let name = id.name in
+  let open String in
+  List.map
+    ~f:(fun n -> {name= n; id_loc= id.id_loc})
+    ( if name = "multiply_log" || name = "binomial_coefficient_log" then [name]
+    else if is_suffix ~suffix:"_lpmf" name then
+      [name; drop_suffix name 5 ^ "_lpdf"; drop_suffix name 5 ^ "_log"]
+    else if is_suffix ~suffix:"_lpdf" name then
+      [name; drop_suffix name 5 ^ "_lpmf"; drop_suffix name 5 ^ "_log"]
+    else if is_suffix ~suffix:"_lcdf" name then
+      [name; drop_suffix name 5 ^ "_cdf_log"]
+    else if is_suffix ~suffix:"_lccdf" name then
+      [name; drop_suffix name 6 ^ "_ccdf_log"]
+    else if is_suffix ~suffix:"_cdf_log" name then
+      [name; drop_suffix name 8 ^ "_lcdf"]
+    else if is_suffix ~suffix:"_ccdf_log" name then
+      [name; drop_suffix name 9 ^ "_lccdf"]
+    else if is_suffix ~suffix:"_log" name then
+      [name; drop_suffix name 4 ^ "_lpmf"; drop_suffix name 4 ^ "_lpdf"]
+    else [name] )
```

Comment:
[Another holdover](https://github.com/stan-dev/stanc3/blob/763421b9493ea1f7f7446392176435d85055afdb/src/frontend/Semantic_check.ml#L92). I agree - keeping all these helper functions more or less unchanged was useful for a lot of the bulk changes I was making, but I will take a closer look at them now.

For this function specifically, I agree doing a right split on `_` would be better.

Reply:
Actually, `_cdf_log` and `_ccdf_log` break this, because we'd need to match on the second to last underscore. Not sure if there is a cleaner way than the big if/else because of this

Reply:
It's a shame, because we could _almost_ do this

```ocaml
    if name = "multiply_log" || name = "binomial_coefficient_log" then [name]
    else
      match rsplit2 ~on:'_' name with
      | Some (stem, "lpmf") | Some (stem, "lpdf") | Some (stem, "log") ->
          [stem ^ "_lpmf"; stem ^ "_lpdf"; stem ^ "_log"]
      | Some (stem, "lcdf") | Some (stem, "cdf_log") ->
          [stem ^ "_lcdf"; stem ^ "_cdf_log"]
      | Some (stem, "lccdf") | Some (stem, "ccdf_log") ->
          [stem ^ "_lccdf"; stem ^ "_ccdf_log"]
      | _ -> [name]
```

but all the cdf_logs will end up in the first case, and I don't know of any obvious way around this